src/combinatorics/graph/basic.lean

```diff
@@ -0,0 +1,620 @@
+
+
+/-! ### Sort world -/
+
+-- insert `quiver` definition here
+-- Note, `quiver` itself is a class.
+-- Note, we could generalize `quiver α` to `quiver α β`
+
+/-- `quiver_class F α` states that `F` is a type of quivers on `α`.
+
+Extend this class when you extend `quiver`. -/
+class quiver_class (F : Type*) (α : out_param $ Type*) :=
+(hom : F → α → α → Sort*)
+
+instance quiver.quiver_class : quiver_class (quiver α) α :=
+⟨@quiver.hom α⟩
+
+/-- `a ⟶[G] b` is the type of edges from `a` to `b` in `G`. -/
+@[nolint has_inhabited_instance] def graph.hom [quiver_class F α] : F → α → α → Sort* := quiver_class.hom
+
+notation a ` ⟶[`:25 G `] ` b := graph.hom G a b
+
+-- We might need to replace `u₃` by `u₃ + 1` here to avoid providing an adjacency relation for stuff
+-- that already has one, ie when `hom` is already Prop-valued
+@[priority 100] -- see Note [lower instance priority]
+instance quiver_class.to_rel_class (F : Type u₁) (α : Type u₂) [quiver_class.{u₁ u₂ u₃} F α] :
+  rel_class F α α :=
+⟨λ f a b, nonempty (a ⟶[f] b)⟩
+
+instance : has_bot (quiver α) := ⟨⟨λ _ _, pempty⟩⟩
+instance : inhabited (quiver α) := ⟨⊥⟩
+
+
+/-- The type of multigraphs on `α`. -/
+structure multigraph (α : Type u₂) extends quiver.{u₁ u₂} α : Type (max u₁ u₂) :=
+(hom_fun (a b : α) : (a ⟶ b) → (b ⟶ a))
+(hom_fun_comp (a b : α) : hom_fun a b ∘ hom_fun b a = id)
+
+/-- `multigraph_class F α` states that `F` is a type of multigraphs on `α`.
+
+Extend this class when you extend `multigraph`. -/
+class multigraph_class (F : Type u₁) (α : out_param $ Type u₂)
+  extends quiver_class.{u₁ u₂ u₃} F α : Type (max u₁ u₂ u₃) :=
+(hom_fun (f : F) (a b : α) : (a ⟶[f] b) → (b ⟶[f] a))
+(hom_fun_comp (f : F) (a b : α) : hom_fun f a b ∘ hom_fun f b a = id)
+
+instance multigraph.multigraph_class : multigraph_class (multigraph α) α :=
+{ hom := λ G, G.hom,
+  hom_fun := λ G, G.hom_fun,
+  hom_fun_comp := λ G, G.hom_fun_comp }
+
+/-- Turns an edge around. -/
+protected def graph.hom.symm [multigraph_class F α] {f : F} {a b : α} : (a ⟶[f] b) → (b ⟶[f] a) :=
+multigraph_class.hom_fun _ _ _
+
+lemma graph.hom_symm_symm [multigraph_class F α] {f : F} {a b : α} (e : a ⟶[f] b) :
+  e.symm.symm = e :=
+begin
+  change _ = id e,
+  rw ←multigraph_class.hom_fun_comp f b a,
+  refl,
+end
+
+-- `dangerous_instance` does not know that `W` is used only as an `out_param`.
+@[priority 100, nolint dangerous_instance] -- see Note [lower instance priority]
+instance multigraph_class.to_symm_rel_class [multigraph_class F α] : symm_rel_class F α :=
+{ to_fun := graph.adj,
+  to_fun_symm := λ f a b h, h.map graph.hom.symm }
+
+instance : has_bot (multigraph α) :=
+⟨{ hom := λ _ _, pempty,
+  hom_fun := λ _ _, id,
+  hom_fun_comp := λ _ _, rfl }⟩
+
+instance : inhabited (multigraph α) := ⟨⊥⟩
+
+
+/-- The type of simple quivers on `α`, aka simple multidigraphs. -/
+structure simple_quiver (α : Type*) extends quiver α :=
+(hom_self (a : α) : is_empty (a ⟶ a))
```

Comment:
I don't know how useful this structure is, maybe this could just be a `Prop`? Also I'm not sure about the terminology "simple" here, is that standard? What about something like 
```
class loopless (α : Type*) [quiver αextends] : Prop :=
(hom_self (a : α) : is_empty (a ⟶ a)) 
```
Same for `simple_multigraph` and the weighted versions.